dsound_main.c

```diff
@@ -646,7 +646,7 @@ HRESULT get_mmdevice(EDataFlow flow, const GUID *tgt, IMMDevice **device)
         {
             IMMDeviceCollection_Release(coll);
             release_mmdevenum(devenum, init_hr);
-            return DS_OK;
+            return DSERR_NODRIVER;
```

Comment:
This causes the function to always return an error, and leak `*device` when it finds a GUID match (since the caller won't know it wasn't released).

Reply:
Problem is if it's left to return `DS_OK` then the aforementioned issue with wild pointers still occurs; maybe the issue actually lies elsewhere?

Reply:
What wild pointer are you seeing? The `IMMDevice` in `*device` remains valid (unreleased), for the caller to use and release later. The `IMMDeviceEnumerator` and `IMMDeviceCollection` are released right before returning, and not returned themselves anywhere. What's the actual issue you're running into?

Reply:
In my tests what happens is that `init_hr` is marked as success causing the instance inside `devenum` to be released by the call to `CoUninitialize()` within `release_mmdevenum()`. This simultaneously invalidates the instance and frees the memory pointed to by `device` while leaving the pointer itself dangling. In turn `DS_OK` is returned to the calling function (in this case `DSShare_Create()`) which then passes the pointer to `IMMDevice_OpenPropertyStore()` which attempts to access the member variable `lpVtbl`; consequently an access violation/segfault is triggered crashing the whole calling program.

Reply:
So uninitializing COM implicitly destroys the `IMMDevice`. Then, it needs to stay initialized when returning a valid device and have the caller uninitialize it when it's done with the device.

Reply:
Should this change/PR be thrown out then?

Reply:
Yeah, it breaks anything that needs the function to succeed, and would leak in cases where it wouldn't uninitialize COM.